### PR TITLE
Make MSVC use [[nodiscard]]

### PR DESCRIPTION
**Pull request by AlexGuteniev**
Uniformly detect attributes for __cplusplus and _MSVC_LANG

### CODE REVIEW

File: include/fmt/core.h [hunk 1]
```diff
@@ -82,11 +82,17 @@
...
+#  define FMT_CPLUSPLUS __cplusplus
+#endif
+
 #define FMT_HAS_CPP14_ATTRIBUTE(attribute) \
-  (__cplusplus >= 201402L && FMT_HAS_CPP_ATTRIBUTE(attribute))
+  (FMT_CPLUSPLUS >= 201402L && FMT_HAS_CPP_ATTRIBUTE(attribute))
 
 #define FMT_HAS_CPP17_ATTRIBUTE(attribute) \
-  (__cplusplus >= 201703L && FMT_HAS_CPP_ATTRIBUTE(attribute))
+  (FMT_CPLUSPLUS >= 201703L && FMT_HAS_CPP_ATTRIBUTE(attribute))
```
Comment (AlexGuteniev) [inline include/fmt/core.h:95]: I'm not even sure why guard against C++17 attributes in C++14 mode if compiler exposes them unconditionally. Compilers are happy to warn even in C++14 mode: https://godbolt.org/z/TrrKhq355

Comment (vitaut) [reply include/fmt/core.h:95]:
```
<source>:3:3: warning: use of the 'nodiscard' attribute is a C++17 extension [-Wc++17-extensions]
[[nodiscard]] int f();
  ^
```
that's why

File: include/fmt/core.h [hunk 2]
```diff
@@ -82,11 +82,17 @@
 #  define FMT_HAS_CPP_ATTRIBUTE(x) 0
 #endif
 
+#ifdef _MSVC_LANG
+#  define FMT_CPLUSPLUS _MSVC_LANG
+#else
+#  define FMT_CPLUSPLUS __cplusplus
+#endif
```
Comment (vitaut) [inline include/fmt/core.h:89]: What is going on here? Doesn't MSVC use the standard `__cplusplus` macro?

Comment (AlexGuteniev) [reply include/fmt/core.h:89]: [/Zc:__cplusplus](https://docs.microsoft.com/en-us/cpp/build/reference/zc-cplusplus?view=msvc-170) enables updated C++ macro. By default `__cplusplus` is defined to some ancient value.

As I read in MSVC STL chat, the reason for this default behavior is the compatibility with some open-source code, in which higher `__cplusplus` values enable not only legitimate C++ features, but also gcc extensions.

You already had `_MSVC_LANG` for `[[fallthrough]]`, I'm just propagating it a bit further.

File: include/fmt/core.h (hunk 1 above)
Comment (AlexGuteniev) [reply include/fmt/core.h:95]: uh, ok

**Comment by vitaut** [comment]
Thank you

**Pull request merged**